### PR TITLE
Speckle integration

### PR DESCRIPTION
**Pull request by Licini**
If you'd like to test on your machine, remember to pull latest change from `compas_cloud` first.

`COMPAS__init`
`COMPAS__cloud_start`
`COMPAS__load` a file if you want.
[COMPAS.zip](https://github.com/BlockResearchGroup/compas_ui/files/8428884/COMPAS.zip)
View project info (update is temporarily disabled)
`COMPAS__project`

- In the form, Value fields are `stream_id`s of corresponding items.
- For new files the steam_ids are emtpy, they will be filled once the project is pushed once

After making changes
`COMPAS__project_push`
- The stream_id of the project will be kept if you save after first push

pull latest changes
`COMPAS__project_pull`

### CODE REVIEW

**Comment by Licini** [comment]
There are few things to talk about, perhaps let's have a meeting first walking through them

**Comment by Licini** [comment]
@brgcode ping

File: apps/Rhino/COMPAS/dev/config.json [hunk 1]
```diff
@@ -208,7 +208,11 @@
     },
     "settings": {
         "cloud": {
-            "background": true
+            "background": false,
+            "speckle": {
+                "host": "speckle.xyz",
+                "token": "4b6a06f4c7b114e3b4115e1bba5536261cb4d3bf20"
```
Comment (Licini) [inline apps/Rhino/COMPAS/dev/config.json:214]: Just for testing purpose, will get this removed when merging

File: apps/Rhino/COMPAS/dev/COMPAS__project_cmd.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+__commandname__ = "COMPAS__project"
+
+
+def RunCommand(is_interactive):
+    from compas_ui.ui import UI
+
+    ui = UI()
+    ui.update_project()
+    # ui.record()
```
Comment (tomvanmele) [inline apps/Rhino/COMPAS/dev/COMPAS__project_cmd.py:9]: without recording there is no undo/redo point associated with this

File: src/compas_ui/project.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    @property
+    def ui(self):
+        if not hasattr(self, '_ui') or self._ui is None:
+            from compas_ui.ui import UI
+            self._ui = UI()
+        return self._ui
+
+    @property
+    def state(self):
+
```
Comment (tomvanmele) [inline src/compas_ui/project.py:30]: `black` will remove this empty line...

File: src/compas_ui/rhino/forms/scene.py [hunk 4]
```diff
@@ -50,6 +51,39 @@ def on_click(sender, e):
         return control
 
 
+class PushCell(Eto.Forms.CustomCell):
```
Comment (Licini) [inline src/compas_ui/rhino/forms/scene.py:54]: Parts about forms are outdated, will get this updated once we finalize the `Project` class

File: src/compas_ui/rhino/forms/commit.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
+from __future__ import print_function
```
Comment (Licini) [inline src/compas_ui/rhino/forms/commit.py:1]: Same... will update this once we finalize `Project`

File: src/compas_ui/project.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                if hasattr(data, 'stream_id'):
+                    data_dict[guid] = data.stream_id
+                else:
+                    data_dict[guid] = None
+
+        return {
+            "name": self.name,
+            "stream_id": self.stream_id,
+            "session": self.ui.session.data,
+            "data": list(data_dict.values()),
```
Comment (tomvanmele) [inline src/compas_ui/project.py]: why not just use the dict?

File: src/compas_ui/project.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+            "objects": objects,
+            "scenes": [self.ui.scene.stream_id]
+        }
+
+    @state.setter
+    def state(self, state):
+        self.name = state['name']
+        self.stream_id = state['stream_id']
```
Comment (tomvanmele) [inline src/compas_ui/project.py]: this seems to not recreate the state of the object

Comment (tomvanmele) [reply src/compas_ui/project.py]: regardless of the speckle part, getting and setting state should be compatible and reciprocal

File: src/compas_ui/project.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+            data.stream_id = self.ui.proxy.speckle_push(stream_id=data.stream_id, item=data, name=data.name+".data", message=message)
+            print("data", data.stream_id)
+
+        for obj in self.ui.scene.objects:
+            # Push object state
+            obj_state = obj.state
+            for key in list(obj_state.keys()):
+                if key.startswith('_guid'):
+                    del obj_state[key]
+            del obj_state['_item']
```
Comment (tomvanmele) [inline src/compas_ui/project.py]: maybe this can be grouped i a `streamdata` property on the object

Comment (tomvanmele) [reply src/compas_ui/project.py]: or `stream` which is a dict with `"id"` and `"state"` or `"content"` as keys

File: src/compas_ui/project.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        scene = self.ui.proxy.speckle_pull(state['scenes'][0])
+        self.ui.scene.settings = scene["settings"]
+        for obj_stream_id in scene['objects']:
+            # Load object state
+            obj_state = self.ui.proxy.speckle_pull(obj_stream_id)
+            # Load object data
+            data = self.ui.proxy.speckle_pull(obj_state['data'])
+            obj = self.ui.scene.add(data)
+            del obj_state['data']
+            obj.state = obj_state
```
Comment (tomvanmele) [inline src/compas_ui/project.py]: this should be handled by setting `self.state`

**Comment by Licini** [comment]
How about now? @brgcode , several things:
- Using `dict` of `{ guid: wrapper/metadata }` for each project attributes
- For the wrapper (or stream or metadata whichever we like to call it), it contains `stream_id`, `state` and `_pointer` which points to the real-time python object in memory. These pointers are excluded for pickling and speckle. 
- I isolated out an `update` function to be explicitly called for getting latest states of scenes and data. The `_pointer`s are attached here the first time this function is called.
- IMO, all network operation should only happen in `push` and `pull` functions, they should not happen when we hit redo or undo. I'd consider state of project merely as collective "snapshot" of states of scene, data etc, so get and set state here only affect the "snapshot" itself, the impact to actual scene and data happens when push and pull are called. 

Let me know what you think

File: src/compas_ui/project.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        """Update the project state base on what's available in the scene."""
+
+        self.data.clear()
+        for obj in self.ui.scene.objects:
+            # Update data states
+            if str(obj.item.guid) not in self.data:
+                stream_id = None
+                if hasattr(obj.item, 'stream_id'):
+                    stream_id = obj.item.stream_id
+                item_state = json.loads(compas.json_dumps(obj.item))
```
Comment (Licini) [inline src/compas_ui/project.py:38]: is there a more elegant api for converting compas.data to json serializable?

File: src/compas_ui/project.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        self.state = state
+
+    @property
+    def ui(self):
+        if not hasattr(self, '_ui') or self._ui is None:
+            from compas_ui.ui import UI
+            self._ui = UI()
+        return self._ui
+
+    def update(self):
```
Comment (Licini) [inline src/compas_ui/project.py:28]: perhaps `capture` or `snapshot` is better name for this function..

**Comment by Licini** [comment]
This will go into its independent repo

**Pull request closed without merging**